### PR TITLE
#16 Set Server URL via QR code

### LINKED ISSUE #16: Set Server URL via QR code
To setup quickly a server url on the first application launch, user should be able to set the url automatically via QR code using the camera inside the application.

### PR DESCRIPTION
**Pull request by gfusee**
- Added QRCodeViewController wrapped in a SwiftUI view.
- Auto save changes if QR code is valid (start with "https://...").
- Added camera permission request in Info.plist and error message if camera access is rejected. 
- Added a qr code button in the settings view.
- Added two icons for the qr code button (one for each colour scheme).

My English is not perfect, feel free to edit messages.

### CODE REVIEW

**Comment by gfusee** [comment]
oh master branch changed and there is conflict now... I will correct my pull request soon

**Comment by nesimtunc** [comment]
Hi @Quentus123 thanks a lot! That's a quite good work! Appreciated. Due to license changing we need a little bit time to merge. Just for your information. So, I might need to change the status of the pull request to _Draft_ in to prevent accepting the merge.

**Comment by gfusee** [comment]
Hi @nesimtunc, no prob!

File: CoThings/Views/Settings/QRCodeView.swift [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+//
+
+import Foundation
+import SwiftUI
+import AVFoundation
+import UIKit
+
+struct QRCodeView: UIViewControllerRepresentable {
+    
+    private var controller: QRCodeViewController = QRCodeViewController()
```
Comment (umurgdk) [inline CoThings/Views/Settings/QRCodeView.swift:16]: Do not create the view controller in here. It's life time is controlled by the swiftui framework so you should create the ui view controller whenever swiftui asks you.

File: CoThings/Views/Settings/QRCodeView.swift [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+    
+    init(onFoundCode: @escaping (String) -> Void, onCameraError: @escaping () -> Void) {
+        controller.foundQRCode = onFoundCode
+        controller.failedToOpenCamera = onCameraError
+    }
+    
+    func makeUIViewController(context: Context) -> QRCodeViewController {
+        return controller
```
Comment (umurgdk) [inline CoThings/Views/Settings/QRCodeView.swift:28]: A new `QRViewController` should be created here.

File: CoThings/Views/Settings/ServerSettingsView.swift [hunk 3]
```diff
@@ -27,26 +30,65 @@ struct ServerSettingsView: View {
...
+                        self.isScanningQRCode = false
+                    }))
             }
-            
-            Section {
-                Button("Done", action: self.save)
+            Form {
+                Section(header: Text("Server URL")) {
+                    GeometryReader { metrics in
+                        HStack {
```
Comment (umurgdk) [inline CoThings/Views/Settings/ServerSettingsView.swift]: There is no need for two nested HStacks here, you can add more views to the inner HStack

File: CoThings/Views/Settings/ServerSettingsView.swift [hunk 4]
```diff
@@ -27,26 +30,65 @@ struct ServerSettingsView: View {
...
+                                    .keyboardType(.URL)
+                                    .disableAutocorrection(true)
+                                    .autocapitalization(.none)
+                                    .frame(maxWidth: .infinity)
+                                
+                            }
+                            Image(self.colorScheme == .dark ? "qrWhiteIcon": "qrBlackIcon")
+                            .resizable()
+                                .frame(width: metrics.size.height, height: metrics.size.height, alignment: .center)
+                            .onTapGesture {
```
Comment (umurgdk) [inline CoThings/Views/Settings/ServerSettingsView.swift]: Using a Button instead of an Image with tap gestures might explain the intention better, if you don't want Button to turn the image into blue accent color please choose `Render As: Original Image` in the asset catalog. But it looks better in blue anyways.

```swift
Button(action: { ... }) {
    Image(...)
      .resizable()
}

File: CoThings/Views/Settings/ServerSettingsView.swift [hunk 5]
```diff
@@ -27,26 +30,65 @@ struct ServerSettingsView: View {
...
 	}
     
+    private func parseQRCode(code: String) {
+        if(code.hasPrefix("https://")) {
+            serverHostname = String(code.dropFirst("https://".count))
+        } else {
+            serverHostname = code
+        }
+        DispatchQueue.main.asyncAfter(deadline: .now() + 0.3) {
+            self.save()
```
Comment (umurgdk) [inline CoThings/Views/Settings/ServerSettingsView.swift]: Having server hostname is populated by a QR code has a different experience than where the user writes the url into text field. `self.save()` dismisses this view, so user will not know what kind of url was entered. It might be okay to run `stateController.saveConfiguration` here but dismissing this view might confuse the user.

File: CoThings/Views/Settings/QRCodeView.swift (hunk 1 above)
Comment (umurgdk) [inline CoThings/Views/Settings/QRCodeView.swift]: ViewController is owned by the swiftui framework, so creating `QRCodeViewController()` here might cause problems, since `UIViewControllerRepresentable` is expected to create a new instance when `makeUIViewController` called.

File: CoThings/Views/Settings/ServerSettingsView.swift [hunk 6]
```diff
@@ -27,26 +30,65 @@ struct ServerSettingsView: View {
...
-                HStack(alignment: .firstTextBaseline, spacing: 1) {
-                    Text("https://")
-                    TextField("demo-eu.cothings.app", text: $serverHostname)
-                        .keyboardType(.URL)
-                        .disableAutocorrection(true)
-                        .autocapitalization(.none)
-                        .frame(maxWidth: .infinity)
-                }
+        ZStack{
+            NavigationLink("Scan QR Code", destination: QRCodeView(
```
Comment (umurgdk) [inline CoThings/Views/Settings/ServerSettingsView.swift]: What do you think about presenting Scan QR Code screen as a modal?

File: CoThings/Views/Settings/QRCodeView.swift [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+
+extension QRCodeViewController: AVCaptureMetadataOutputObjectsDelegate {
+    func metadataOutput(_ output: AVCaptureMetadataOutput, didOutput metadataObjects: [AVMetadataObject], from connection: AVCaptureConnection) {
+        captureSession.stopRunning()
+
+        if let metadataObject = metadataObjects.first {
+            guard let readableObject = metadataObject as? AVMetadataMachineReadableCodeObject else { return }
+            guard let stringValue = readableObject.stringValue else { return }
+            AudioServicesPlaySystemSound(SystemSoundID(kSystemSoundID_Vibrate))
+            foundQRCode(stringValue)
```
Comment (umurgdk) [inline CoThings/Views/Settings/QRCodeView.swift]: It might be better to check if the `stringValue` is a proper URL before calling `foundQRCode` callback. Accepting non URL strings might confuse the user since the QRCodeViewController is automatically dismissed on QR code found.

File: CoThings/Views/Settings/ServerSettingsView.swift [hunk 8]
```diff
@@ -27,26 +30,65 @@ struct ServerSettingsView: View {
...
+                    Alert(title: Text("Error"), message: Text("Cannot open camera.\nMake sure to allow CoThings to access your camera."), dismissButton: .default(Text("Back"), action: {
+                        self.isScanningQRCode = false
+                    }))
             }
-            
-            Section {
-                Button("Done", action: self.save)
+            Form {
+                Section(header: Text("Server URL")) {
+                    GeometryReader { metrics in
```
Comment (umurgdk) [inline CoThings/Views/Settings/ServerSettingsView.swift]: No need for a geometry reader, please use `layoutPriority`'s for text field and label.

File: CoThings/Views/Settings/QRCodeView.swift [hunk 9]
```diff
@@ -0,0 +1,115 @@
...
+        view.layer.addSublayer(previewLayer)
+
+        captureSession.startRunning()
+    }
+
+    override func viewWillAppear(_ animated: Bool) {
+        super.viewWillAppear(animated)
+
+        if (captureSession?.isRunning == false) {
+            captureSession.startRunning()
```
Comment (umurgdk) [inline CoThings/Views/Settings/QRCodeView.swift]: Documentation warns about this is a blocking call and running this method in a separate queue in the background is recommended.

**Review by umurgdk — CHANGES_REQUESTED** [review]

File: CoThings/Views/Settings/ServerSettingsView.swift (hunk 3 above)
Comment (gfusee) [reply CoThings/Views/Settings/ServerSettingsView.swift]: I used two nested HStack to correct alignement issue :
100085075_1316790851853343_4426300893331193856_n

To fix this issue with one HStack I changed the first Text into a disabled TextField : 
```
HStack(spacing: 1) {
    TextField("", text: .constant("https://"))
        .disabled(true)
        .fixedSize()
    TextField("demo-eu.cothings.app", text: self.$serverHostname)
        .keyboardType(.URL)
        .disableAutocorrection(true)
        .autocapitalization(.none)
        .frame(maxWidth: .infinity)
        .layoutPriority(1)
    Button(action: {
        self.isScanningQRCode = true
    }) {
        Image(self.colorScheme == .dark ? "qrWhiteIcon": "qrBlackIcon")
            .resizable()
            .frame(minWidth: 35)
            .scaledToFit()
    }
    
}
```

File: CoThings/Views/Settings/ServerSettingsView.swift (hunk 6 above)
Comment (gfusee) [reply CoThings/Views/Settings/ServerSettingsView.swift]: Oh yes this will be much better! In order to do this I added a title and a dismiss button in the controller : 
99289016_180122119973758_9107629862297796608_n

Comment (umurgdk) [reply CoThings/Views/Settings/ServerSettingsView.swift]: This looks great!

File: CoThings/Views/Settings/QRCodeView.swift [hunk 10]
```diff
@@ -74,25 +80,64 @@ class QRCodeViewController: UIViewController {
...
+
+        sessionQueue.async {
+            self.captureSession.startRunning()
+        }
     }
 
     override func viewWillAppear(_ animated: Bool) {
         super.viewWillAppear(animated)
 
         if (captureSession?.isRunning == false) {
```
Comment (umurgdk) [inline CoThings/Views/Settings/QRCodeView.swift]: Checking the equality of a boolean feels weird, instead of it `!captureSession?.isRunning` feels more natural.

File: CoThings/Views/Settings/QRCodeView.swift [hunk 11]
```diff
@@ -74,25 +80,64 @@ class QRCodeViewController: UIViewController {
...
+            sessionQueue.async {
+                self.captureSession.startRunning()
+            }
         }
     }
 
     override func viewWillDisappear(_ animated: Bool) {
         super.viewWillDisappear(animated)
 
         if (captureSession?.isRunning == true) {
```
Comment (umurgdk) [inline CoThings/Views/Settings/QRCodeView.swift]: Same here there is no need to check if a boolean is equal to true. `if (captureSession?.isRunning)` is enough

**Review by umurgdk — CHANGES_REQUESTED** [review]
Thank so much for the changes!

File: CoThings/Views/Settings/QRCodeView.swift (hunk 11 above)
Comment (gfusee) [reply CoThings/Views/Settings/QRCodeView.swift]: totally agree! I didn't notice that there was '== true', I also removed optional chaining since self.captureSession is not optional, thanks!

**Comment by nesimtunc** [comment]
Hi @Quentus123 !

Sorry and thank you for waiting! Finally we've done with license changing. I guess this _Pull Request_ already **done** with _requested changes_; if so you may change the status to `Ready for review` so I can merge but probably there will be some conflicts.

You might need to pull and merge the `master` branch in order to resolve the conflicts, I could do that too but you need to do it to see the license changing, so it means you're aware of the change.  

Basically the change is from **AGPL-3.0** to **GPL-3.0** with some iOS distribution related additions to be seen in [LICENSE](https://github.com/rainlab-inc/cothings-ios/blob/master/LICENSE) and also check out our [Contributing Guidelines](https://github.com/rainlab-inc/cothings-ios/blob/master/CONTRIBUTING.md).

Thanks!

**Comment by gfusee** [comment]
I'll see tomorrow to resolve conflicts!
